Feature/issue 778 add mpi build

wds15 opened the pull request:
#### Submission Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [ ] Run cpplint: `make cpplint`
- [X] Declare copyright holder and open-source license: see below

#### Summary:
Add build for MPI to makefiles

#### Intended Effect:
Trigger automatic build of boost MPI for tests with _mpi_ in their path IF STAN_MPI is defined in the make/local file.

#### How to Verify:
1. Define STAN_MPI in make/local (the value used does not matter).
2. `touch test/unit/fake_mpi_test.cpp`
3. `make test/unit/fake_mpi_test`

This should trigger the automatic build of the boost MPI and its dependencies as dynamic libraries.

#### Side Effects:

#### Documentation:

What we need is a working dynamic library libboost_mpi.dylib and libboost_serialization.dylib (with an .so ending on Linux!). In order to get there we use the boost build system which should be good at figuring out how the MPI installation on a given system needs to be used. If the defaults fail, then the user may himself configure the boost MPI build process by editing the lib/boost/user-config.jam file; the instructions how to do that are found in the boost MPI documentation.

Note that for the matter of the makefiles I am using as target lib/boost/stage/lib/mpi.so, because this files is always generated with the same file ending - on Mac and on Linux.

As I only got MPI working with dynamic linking, there are additional steps on Mac to ensure that the final binary can be executed without additional user interaction. On Mac the absolute paths to the dynamic libraries have to be hardcoded into the final executable. For this to work, the dynamic libraries need to be prepared using then install_name tool. Once that is done, the rpath linker option is sufficient to make things work. On Linux this additional step is not needed.

The plan for cmdstan is to add an if statement to the phony build target. In this if we can then trigger the MPI build whenever STAN_MPI is defined in make/local.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

seantalts commented on make/libraries:
```diff
@@ -65,6 +65,28 @@ LIBIDAS = $(IDAS)/lib/libsundials_nvecserial.a $(IDAS)/lib/libsundials_idas.a
+	cd $(BOOST); ./bootstrap.sh
+	cd $(BOOST); echo "using mpi ;" >> user-config.jam
+	cd $(BOOST); ./b2 --user-config=user-config.jam --with-mpi --with-serialization
+	@if [ "$(OS_TYPE)" = "mac" ]; then\
```
So what follows here (the install_name_tool commands) are commands that need to be run on a Mac when we're building `mpi.so`, but not necessarily when we're compiling code against `mpi.so`?

If this is correct I think this structure makes sense. But if it were the case that the `install_name_tool` commands mentioned were needed any time you were on a Mac and had `STAN_MPI` defined, I'd be inclined to put them in the `os_mac` file.

wds15 replied:
It took me while starring at the documentation of macOS linker tools and stuff on the net to figure this out. I never intended to dive this deep into it, but now it works. Should we collect more dynamic libraries in stan-math which we want to distribute, we may revisit as to how we set the rpath logic.

seantalts commented on make/os_linux:
```diff
@@ -22,3 +22,9 @@ ifeq (clang++,$(CC_TYPE))
+  LDFLAGS += -L$(BOOST)/stage/lib -lboost_serialization -lboost_mpi -Wl,-rpath,$(BOOST)/stage/lib
+  CXXFLAGS += -DSTAN_MPI
+  GTEST_CXXFLAGS += -DSTAN_MPI
+  LIBMPI=$(BOOST)/stage/lib/mpi.so
```
Right - was wondering why you thought it needed an immediate set here? Are there issues with the lazy set or does it take longer or something? Do people generally prefer immediate unless they need the laziness?

syclik replied:
Yes, there are issues with the lazy set. Yes, it does take longer, but that's a theoretic concern here. I don't think the time is going to matter much.

What was really bothering me here was the use of the variable. When it's set, it's used as a dependency. When it's not set, the absence of the expansion is being used as a lack of a dependency. That's where it's tricky and it took me 3-4 passes of this code to figure out. That's what I'm trying to cut out. We don't need to rely on make features that allow you to lazy evaluate and make it hard to reason about the state of things.

seantalts commented on make/tests:
```diff
@@ -60,6 +60,15 @@ IDAS_TESTS := $(subst .cpp,$(EXE),$(shell find test -name *idas*_test.cpp) $(she
+##
+
+MPI_TESTS := $(subst .cpp,$(EXE),$(shell find test -name *_mpi_*test.cpp))
+$(MPI_TESTS) : $(LIBMPI)
```
👍  It will also be set for CmdStan, which I think is correct.

wds15 replied:
So I leave the `CXXFLAGS` definition where it was, right? To me that is the right thing as we want these things to be defined not only for the tests whenever MPI is turned on (cmdstan can then pickup these definitions).

wds15 commented on make/libraries:
```diff
@@ -65,6 +65,28 @@ LIBIDAS = $(IDAS)/lib/libsundials_nvecserial.a $(IDAS)/lib/libsundials_idas.a
 .PHONY: clean-libraries
 clean-libraries:
 	$(RM) $(sort $(SUNDIALS_CVODES) $(SUNDIALS_NVECSERIAL) $(LIBCVODES) $(SUNDIALS_IDAS) $(SUNDIALS_IDAS_NVECSERIAL) $(LIBIDAS))
+	[ -d "$(BOOST)/stage/lib" ] && $(RM) $(BOOST)/stage/lib/*  
```
OK

wds15 commented on make/os_linux:
```diff
@@ -22,3 +22,9 @@ ifeq (clang++,$(CC_TYPE))
+ifdef STAN_MPI
+  LDFLAGS += -L$(BOOST)/stage/lib -lboost_serialization -lboost_mpi -Wl,-rpath,$(BOOST)/stage/lib
+  CXXFLAGS += -DSTAN_MPI
+  GTEST_CXXFLAGS += -DSTAN_MPI
```
... because for tests we anyway add the CXXFLAGS? If so, then sure, it's not needed.

wds15 commented:
I can look into the Jenkinsfile, but I have to say that I have never dealt with the config of Jenkins. Maybe I need help, let's see. For the testing there is actually an additional complication which is that we have to run the mpi tests through `mpirun -np 2` (for using 2 cores during testing). So a call to

`./tests/unit/some_mpi_test`

becomes

`mpirun -np 2 ./tests/unit/some_mpi_test`

If no MPI stuff is done in the test (so a non-MPI test), then this won't hurt or do anything. So this would need to be configured either in the Jenkinsfile (if possible) OR we update the runTests.py script to evoke `mpirun` for all tests with `mpi_` in their file name. Any preference?

I wasn't aware to do a tricky thing with `LIBMPI` - I just tried to mimick the other libraries logic here. However, what took me a long time to realize is that stupid ORDER matters for makefiles (the `libraries` make is included very early on, even before the OS).

@syclik I hope got all change requests, will address them later.

seantalts commented:
I like the idea of updating runTests to switch to mpirun for the mpi tests.
I can help with that and the Jenkinsfile. Do we want to add a separate
section for MPI enabled unit tests, is that it?
On Thu, Mar 8, 2018 at 03:14 wds15 <notifications@github.com> wrote:

> I can look into the Jenkinsfile, but I have to say that I have never dealt
> with the config of Jenkins. Maybe I need help, let's see. For the testing
> there is actually an additional complication which is that we have to run
> the mpi tests through mpirun -np 2 (for using 2 cores during testing). So
> a call to
>
> ./tests/unit/some_mpi_test
>
> becomes
>
> mpirun -np 2 ./tests/unit/some_mpi_test
>
> If no MPI stuff is done in the test (so a non-MPI test), then this won't
> hurt or do anything. So this would need to be configured either in the
> Jenkinsfile (if possible) OR we update the runTests.py script to evoke
> mpirun for all tests with mpi_ in their file name. Any preference?
>
> I wasn't aware to do a tricky thing with LIBMPI - I just tried to mimick
> the other libraries logic here. However, what took me a long time to
> realize is that stupid ORDER matters for makefiles (the libraries make is
> included very early on, even before the OS).
>
> @syclik <https://github.com/syclik> I hope got all change requests, will
> address them later.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-371412502>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7CGw6BkEj5YEMfnLxZSKdrM4PihWks5tcOhEgaJpZM4Sa35S>
> .
>

syclik commented:
We want a separate job where all the unit tests are run with the MPI stuff
turned on. Not just the MPI tests.

On Thu, Mar 8, 2018 at 7:47 AM, seantalts <notifications@github.com> wrote:

> I like the idea of updating runTests to switch to mpirun for the mpi tests.
> I can help with that and the Jenkinsfile. Do we want to add a separate
> section for MPI enabled unit tests, is that it?
> On Thu, Mar 8, 2018 at 03:14 wds15 <notifications@github.com> wrote:
>
> > I can look into the Jenkinsfile, but I have to say that I have never
> dealt
> > with the config of Jenkins. Maybe I need help, let's see. For the testing
> > there is actually an additional complication which is that we have to run
> > the mpi tests through mpirun -np 2 (for using 2 cores during testing). So
> > a call to
> >
> > ./tests/unit/some_mpi_test
> >
> > becomes
> >
> > mpirun -np 2 ./tests/unit/some_mpi_test
> >
> > If no MPI stuff is done in the test (so a non-MPI test), then this won't
> > hurt or do anything. So this would need to be configured either in the
> > Jenkinsfile (if possible) OR we update the runTests.py script to evoke
> > mpirun for all tests with mpi_ in their file name. Any preference?
> >
> > I wasn't aware to do a tricky thing with LIBMPI - I just tried to mimick
> > the other libraries logic here. However, what took me a long time to
> > realize is that stupid ORDER matters for makefiles (the libraries make is
> > included very early on, even before the OS).
> >
> > @syclik <https://github.com/syclik> I hope got all change requests, will
> > address them later.
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/stan-dev/math/pull/783#issuecomment-371412502>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAxJ7CGw6BkEj5YEMfnLxZSKdrM4PihWks5tcOhEgaJpZM4Sa35S>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-371477302>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_FwPT89ldU2mLLSVhJ8Hwis55at9Mks5tcSh1gaJpZM4Sa35S>
> .
>

seantalts commented:
Kk, sounds good. Will work on that later today.
On Thu, Mar 8, 2018 at 08:38 Daniel Lee <notifications@github.com> wrote:

> We want a separate job where all the unit tests are run with the MPI stuff
> turned on. Not just the MPI tests.
>
> On Thu, Mar 8, 2018 at 7:47 AM, seantalts <notifications@github.com>
> wrote:
>
> > I like the idea of updating runTests to switch to mpirun for the mpi
> tests.
> > I can help with that and the Jenkinsfile. Do we want to add a separate
> > section for MPI enabled unit tests, is that it?
> > On Thu, Mar 8, 2018 at 03:14 wds15 <notifications@github.com> wrote:
> >
> > > I can look into the Jenkinsfile, but I have to say that I have never
> > dealt
> > > with the config of Jenkins. Maybe I need help, let's see. For the
> testing
> > > there is actually an additional complication which is that we have to
> run
> > > the mpi tests through mpirun -np 2 (for using 2 cores during testing).
> So
> > > a call to
> > >
> > > ./tests/unit/some_mpi_test
> > >
> > > becomes
> > >
> > > mpirun -np 2 ./tests/unit/some_mpi_test
> > >
> > > If no MPI stuff is done in the test (so a non-MPI test), then this
> won't
> > > hurt or do anything. So this would need to be configured either in the
> > > Jenkinsfile (if possible) OR we update the runTests.py script to evoke
> > > mpirun for all tests with mpi_ in their file name. Any preference?
> > >
> > > I wasn't aware to do a tricky thing with LIBMPI - I just tried to
> mimick
> > > the other libraries logic here. However, what took me a long time to
> > > realize is that stupid ORDER matters for makefiles (the libraries make
> is
> > > included very early on, even before the OS).
> > >
> > > @syclik <https://github.com/syclik> I hope got all change requests,
> will
> > > address them later.
> > >
> > > —
> > > You are receiving this because your review was requested.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/stan-dev/math/pull/783#issuecomment-371412502>, or
> > mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > AAxJ7CGw6BkEj5YEMfnLxZSKdrM4PihWks5tcOhEgaJpZM4Sa35S>
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/stan-dev/math/pull/783#issuecomment-371477302>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAZ_FwPT89ldU2mLLSVhJ8Hwis55at9Mks5tcSh1gaJpZM4Sa35S
> >
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-371488535>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7IiIGHTmp5bDbpS9V6PtBQq9lNT2ks5tcTQ2gaJpZM4Sa35S>
> .
>

syclik commented:
Thanks! Does that make sense? We want to make sure that the unit tests all
run under both MPI and non MPI. We'd want to catch it as soon as anything
broke under either condition.

On Thu, Mar 8, 2018 at 8:51 AM, seantalts <notifications@github.com> wrote:

> Kk, sounds good. Will work on that later today.
>
> On Thu, Mar 8, 2018 at 08:38 Daniel Lee <notifications@github.com> wrote:
>
> > We want a separate job where all the unit tests are run with the MPI
> stuff
> > turned on. Not just the MPI tests.
> >
> > On Thu, Mar 8, 2018 at 7:47 AM, seantalts <notifications@github.com>
> > wrote:
> >
> > > I like the idea of updating runTests to switch to mpirun for the mpi
> > tests.
> > > I can help with that and the Jenkinsfile. Do we want to add a separate
> > > section for MPI enabled unit tests, is that it?
> > > On Thu, Mar 8, 2018 at 03:14 wds15 <notifications@github.com> wrote:
> > >
> > > > I can look into the Jenkinsfile, but I have to say that I have never
> > > dealt
> > > > with the config of Jenkins. Maybe I need help, let's see. For the
> > testing
> > > > there is actually an additional complication which is that we have to
> > run
> > > > the mpi tests through mpirun -np 2 (for using 2 cores during
> testing).
> > So
> > > > a call to
> > > >
> > > > ./tests/unit/some_mpi_test
> > > >
> > > > becomes
> > > >
> > > > mpirun -np 2 ./tests/unit/some_mpi_test
> > > >
> > > > If no MPI stuff is done in the test (so a non-MPI test), then this
> > won't
> > > > hurt or do anything. So this would need to be configured either in
> the
> > > > Jenkinsfile (if possible) OR we update the runTests.py script to
> evoke
> > > > mpirun for all tests with mpi_ in their file name. Any preference?
> > > >
> > > > I wasn't aware to do a tricky thing with LIBMPI - I just tried to
> > mimick
> > > > the other libraries logic here. However, what took me a long time to
> > > > realize is that stupid ORDER matters for makefiles (the libraries
> make
> > is
> > > > included very early on, even before the OS).
> > > >
> > > > @syclik <https://github.com/syclik> I hope got all change requests,
> > will
> > > > address them later.
> > > >
> > > > —
> > > > You are receiving this because your review was requested.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/stan-dev/math/pull/783#issuecomment-371412502>,
> or
> > > mute
> > > > the thread
> > > > <https://github.com/notifications/unsubscribe-auth/
> > > AAxJ7CGw6BkEj5YEMfnLxZSKdrM4PihWks5tcOhEgaJpZM4Sa35S>
> > > > .
> > > >
> > >
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/stan-dev/math/pull/783#issuecomment-371477302>, or
> > mute
> > > the thread
> > > <
> > https://github.com/notifications/unsubscribe-auth/AAZ_
> FwPT89ldU2mLLSVhJ8Hwis55at9Mks5tcSh1gaJpZM4Sa35S
> > >
> > > .
> > >
> >
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/stan-dev/math/pull/783#issuecomment-371488535>, or
> mute
> > the thread
> > <https://github.com/notifications/unsubscribe-auth/
> AAxJ7IiIGHTmp5bDbpS9V6PtBQq9lNT2ks5tcTQ2gaJpZM4Sa35S>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-371492015>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_Fy89W9kazR0JdsXNiD8ZqynO72xnks5tcTdagaJpZM4Sa35S>
> .
>

seantalts commented:
Yep.
On Thu, Mar 8, 2018 at 08:57 Daniel Lee <notifications@github.com> wrote:

> Thanks! Does that make sense? We want to make sure that the unit tests all
> run under both MPI and non MPI. We'd want to catch it as soon as anything
> broke under either condition.
>
> On Thu, Mar 8, 2018 at 8:51 AM, seantalts <notifications@github.com>
> wrote:
>
> > Kk, sounds good. Will work on that later today.
> >
> > On Thu, Mar 8, 2018 at 08:38 Daniel Lee <notifications@github.com>
> wrote:
> >
> > > We want a separate job where all the unit tests are run with the MPI
> > stuff
> > > turned on. Not just the MPI tests.
> > >
> > > On Thu, Mar 8, 2018 at 7:47 AM, seantalts <notifications@github.com>
> > > wrote:
> > >
> > > > I like the idea of updating runTests to switch to mpirun for the mpi
> > > tests.
> > > > I can help with that and the Jenkinsfile. Do we want to add a
> separate
> > > > section for MPI enabled unit tests, is that it?
> > > > On Thu, Mar 8, 2018 at 03:14 wds15 <notifications@github.com> wrote:
> > > >
> > > > > I can look into the Jenkinsfile, but I have to say that I have
> never
> > > > dealt
> > > > > with the config of Jenkins. Maybe I need help, let's see. For the
> > > testing
> > > > > there is actually an additional complication which is that we have
> to
> > > run
> > > > > the mpi tests through mpirun -np 2 (for using 2 cores during
> > testing).
> > > So
> > > > > a call to
> > > > >
> > > > > ./tests/unit/some_mpi_test
> > > > >
> > > > > becomes
> > > > >
> > > > > mpirun -np 2 ./tests/unit/some_mpi_test
> > > > >
> > > > > If no MPI stuff is done in the test (so a non-MPI test), then this
> > > won't
> > > > > hurt or do anything. So this would need to be configured either in
> > the
> > > > > Jenkinsfile (if possible) OR we update the runTests.py script to
> > evoke
> > > > > mpirun for all tests with mpi_ in their file name. Any preference?
> > > > >
> > > > > I wasn't aware to do a tricky thing with LIBMPI - I just tried to
> > > mimick
> > > > > the other libraries logic here. However, what took me a long time
> to
> > > > > realize is that stupid ORDER matters for makefiles (the libraries
> > make
> > > is
> > > > > included very early on, even before the OS).
> > > > >
> > > > > @syclik <https://github.com/syclik> I hope got all change
> requests,
> > > will
> > > > > address them later.
> > > > >
> > > > > —
> > > > > You are receiving this because your review was requested.
> > > > > Reply to this email directly, view it on GitHub
> > > > > <https://github.com/stan-dev/math/pull/783#issuecomment-371412502
> >,
> > or
> > > > mute
> > > > > the thread
> > > > > <https://github.com/notifications/unsubscribe-auth/
> > > > AAxJ7CGw6BkEj5YEMfnLxZSKdrM4PihWks5tcOhEgaJpZM4Sa35S>
> > > > > .
> > > > >
> > > >
> > > > —
> > > > You are receiving this because you were mentioned.
> > > > Reply to this email directly, view it on GitHub
> > > > <https://github.com/stan-dev/math/pull/783#issuecomment-371477302>,
> or
> > > mute
> > > > the thread
> > > > <
> > > https://github.com/notifications/unsubscribe-auth/AAZ_
> > FwPT89ldU2mLLSVhJ8Hwis55at9Mks5tcSh1gaJpZM4Sa35S
> > > >
> > > > .
> > > >
> > >
> > > —
> > > You are receiving this because your review was requested.
> > > Reply to this email directly, view it on GitHub
> > > <https://github.com/stan-dev/math/pull/783#issuecomment-371488535>, or
> > mute
> > > the thread
> > > <https://github.com/notifications/unsubscribe-auth/
> > AAxJ7IiIGHTmp5bDbpS9V6PtBQq9lNT2ks5tcTQ2gaJpZM4Sa35S>
> > > .
> > >
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/stan-dev/math/pull/783#issuecomment-371492015>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAZ_Fy89W9kazR0JdsXNiD8ZqynO72xnks5tcTdagaJpZM4Sa35S
> >
> > .
> >
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-371493487>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7EKmv_z1iLPCUXvVOwKwlW4XKKyEks5tcTizgaJpZM4Sa35S>
> .
>

syclik commented:
@wds15, I've got another pull request that makes the build a lot simpler for math. I've been working on that branch for a few months, taking into account this build process and GPU's build process.

One of the reasons I'm trying to keep this pull request clean is because that's how we got into trickiness in the first place. We tend to pile on technical debt and not back it out -- that's just how development works. It's better to fix it when it's spotted instead of later on. Anyway, we're close.

Question: what happens under Windows? Is it not possible or do you just not want to support it right now?

wds15 commented:
@syclik should we merge this pull before your reorg of the make system or the reverse?
I have tried to get it working on Windows for 2-3h and then gave up as it is a mystery to me as to how to install a MPI base system onto it. I would need to try again. For the moment I thought we just error out and that's it for now. Maybe we ask on the forum if some windows expert is able to install MPI (and boost_mpi). The thing is that MPI was built for high-performance computing clusters... which is usually not running Windows.

@seantalts Thanks for looking into this as well!

seantalts commented:
I think MPI is a really high priority for me (Bob mentioned as much as
well). I’d say we get this in as is - Sebastian already put the work into
understanding the current system and creating a solution for it, and Daniel
has in his head the mechanism for translation between the existing system
and his new one, so the labor seems to be well-divided already.

We also all agreed to release first in just CmdStan and Linux I believe in
that call almost a year ago.
On Thu, Mar 8, 2018 at 09:43 wds15 <notifications@github.com> wrote:

> @syclik <https://github.com/syclik> should we merge this pull before your
> reorg of the make system or the reverse?
> I have tried to get it working on Windows for 2-3h and then gave up as it
> is a mystery to me as to how to install a MPI base system onto it. I would
> need to try again. For the moment I thought we just error out and that's it
> for now. Maybe we ask on the forum if some windows expert is able to
> install MPI (and boost_mpi). The thing is that MPI was built for
> high-performance computing clusters... which is usually not running Windows.
>
> @seantalts <https://github.com/seantalts> Thanks for looking into this as
> well!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-371506877>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7Aw9RWK6oOkbYcuo0aB2f0Z3BSzqks5tcUOIgaJpZM4Sa35S>
> .
>

syclik commented:
Definitely no need to wait for the other pull request!

> We also all agreed to release first in just CmdStan and Linux I believe in
that call almost a year ago.

Didn't realize that. (Damn short memory.) I think @wds15 has gotten it going on both mac and linux, so we should test on both? or linux only?

seantalts commented:
I'd probably just leave that up to him, haha. But I think he's been working
primarily with Mac anyway, so seems like we'll get both ;)

On Thu, Mar 8, 2018 at 12:39 PM Daniel Lee <notifications@github.com> wrote:

> Definitely no need to wait for the other pull request!
>
> We also all agreed to release first in just CmdStan and Linux I believe in
> that call almost a year ago.
>
> Didn't realize that. (Damn short memory.) I think @wds15
> <https://github.com/wds15> has gotten it going on both mac and linux, so
> we should test on both? or linux only?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-371563988>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7N3BqCTs0hO2Uvlxuv-Biivjtq84ks5tcWzGgaJpZM4Sa35S>
> .
>

wds15 commented:
These files have already been removed from develop. A merge with develop should solve this. I will do that now.

wds15 commented:
We are going to get Mac and Linux support for MPI. I now remember: For Windows I tried to get the cygwin MPI libraries up and running, but I failed to get that working on my laptop. Since cygwin has openmpi it should be possible to do it, but it was not straightforward for me.

wds15 commented:
Argh... I only tested things to the point that everything got build, but I never executed the `foo_mpi_test` - which I should have done. Anyway, now it works. Changes:

- LIBMPI is not any more defined
- I DO NOT use `mpi.so` as target anymore, because if you do so, then the linker will bring it in and the executable will not start on Mac as the linker tries to get a hold of `mpi.so` which it cannot find. Adding more of this install-whatever magic did not solve this.
- instead I am now making things far more straightforward in terms of makefile logic: In the ifdef block within tests I make the MPI tests dependent on `libboost_mpi$(DLL)` and `libboost_serialization$(DLL)`. The `$(DLL)` is defined os dependent. Moreover, these targets themselves depend on `mpi.so` which is build the usual way. For dylib variants I am moreover adding the install-name magic.

This logic avoids a few if statements and as an upside one even gets an executable which works (yeah).

wds15 commented:
the unit tests with MPI fail on our Jenkins with:
```
[workspace] Running shell script

+ echo STAN_MPI:=true

/Users/Shared/Jenkins/Home/jobs/Math Pipeline/branches/PR-783/workspace@tmp/durable-a5876b4f/script.sh: line 2: /etc/make/local: No such file or directory
```
Not sure how to fix that... maybe @seantalts can have a look?
BTW, there is more than just a `STAN_MPI=true` to add to the `make/local`. In addition the user also must provide the necessary compiler and linker flags such that things get build with reference to the local MPI installation. So for mpich (which is the MPI implementation I would favor to do the testing with) this means to set in `make/local`
```
CXX=mpicxx
STAN_MPI=true
```
Remember that from the Stan side we automate the process to get boost MPI running which is linked against the base MPI installation. For compilation we need to hook up against boost MPI and in order for that library to work you need to throw in things needed to compile and link against the base MPI installation.

My preference for mpich is because our admins here say that it is better integrated with Infiniband (a high-performance network thing).

seantalts commented:
Hah, fixed that /etc/make/local thing. My editor must have inserted that trying to autocomplete the path and I failed to notice.

seantalts commented:
Think I fixed those two points - need to now install mpich on the linux and mac machines in Jenkins.

seantalts commented:
Now that the Jenkins thing is working, it's the code under test that is erroring - we have a bunch of
```
./stan/math/prim/mat/functor/map_rect.hpp:156:2: error: "MPI not yet supported"
#error "MPI not yet supported"
```

wds15 commented:
Ups. Let's just silence the error. I did not foresee that the code gets tested in MPI mode before even any MPI code is there.

wds15 commented:
hmm... now things fail, because of the order how stuff is build. What should solve it is if we do as @syclik suggested and take the `LDFLAGS` and `CXXFLAGS` out of the os files and instead ass to tests this:

```
$(MPI_TESTS): LDFLAGS += -L$(BOOST)/stage/lib -lboost_serialization -lboost_mpi -Wl,-rpath,$(BOOST)/stage/lib
$(MPI_TESTS): CXXFLAGS += -DSTAN_MPI
```

Right now the cvodes libraries get build with the above compiler+linker flags, but the MPI stuff has not yet been build such that it fails.

What I like about having the above definitions in the os file is that this will enable them for all matters for cmdstan. Would it be possible to first run any mpi test which ensures that the boost libraries get build? Or would that be too fragile?

seantalts commented:
Hm, interesting. I bet @syclik's makefile refactor solves this by creating separate flags for the different projects we build. 

I don't think we should rely on the user running tests in the correct order. Do we actually want cvodes to be linked against Boost MPI? If so, can we add a dependency there conditional on `STAN_MPI`? If we don't care if cvodes is built against MPI, maybe we can add that dependency anyway?

(minor note: cvodes probably only gets built with those LDFLAGS because it's not a C++ project (so CXXFLAGS shouldn't apply), but we still need to fix LDFLAGS sharing)

wds15 commented:
We do not need to link anything against boost MPI or the MPI base libraries unless it is the code which actually uses MPI. So neither the cvodes library nor any non-MPI needs to linked against those MPI dependencies (but it does not hurt either).

My latest commit introduces a `CXXFLAGS_MPI` and a `LDFLAGS_MPI` which get defined if `STAN_MPI` is set. Whenever `STAN_MPI` is defined the MPI tests now add the `mpi.so` dependency and for the MPI tests only the `CXXFLAGS` and `LDFLAGS` are modified with their `_MPI` extras.

For cmdstan we should be able to add the `_MPI` additions respectivley; not sure yet, but let‘s go one step after the other. I hope this makes all sense now.

seantalts commented:
Makes sense to me.
On Sun, Mar 11, 2018 at 07:17 wds15 <notifications@github.com> wrote:

> We do not need to link anything against boost MPI or the MPI base
> libraries unless it is the code which actually uses MPI. So neither the
> cvodes library nor any non-MPI needs to linked against those MPI
> dependencies (but it does not hurt either).
>
> My latest commit introduces a CXXFLAGS_MPI and a LDFLAGS_MPI which get
> defined if STAN_MPI is set. Whenever STAN_MPI is defined the MPI tests
> now add the mpi.so dependency and for the MPI tests only the CXXFLAGS and
> LDFLAGS are modified with their _MPI extras.
>
> For cmdstan we should be able to add the _MPI additions respectivley; not
> sure yet, but let‘s go one step after the other. I hope this makes all
> sense now.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-372107470>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7NKj3rw9wz242HefNhpXZUSeHRYrks5tdQfHgaJpZM4Sa35S>
> .
>

syclik commented on Jenkinsfile:
```diff
@@ -147,9 +147,19 @@ pipeline {
+                    steps {
+                        unstash 'MathSetup'
+                        sh "echo CC=mpicxx >> make/local"
+                        sh "echo STAN_MPI:=true >> make/local"
```
we can just leave this as `STAN_MPI=true` (to be consistent with everything else). There's no concern with this being reevaluated multiple times.

seantalts replied:
So when do you use the `:=`? I thought it was whenever you didn't need lazy evaluation.

syclik commented on runTests.py:
```diff
@@ -97,10 +97,12 @@ def makeTest(name, j):
     executable = mungeName(name).replace("/", os.sep)
     xml = mungeName(name).replace(winsfx, "")
     command = '%s --gtest_output="xml:%s.xml"' % (executable, xml)
+    if mpi and "mpi_" in name:
```
This logic should be changed to just `if mpi:`. We want to make sure that when we specify MPI in the build and even if nothing uses MPI, it should still run.

seantalts replied:
This is just saying _how_ to run it - they will all run with MPI compiled in, but some will be run with `mpirun`. Is there some reason we want to run non-mpi unit tests with `mpirun`? (also does that actually work?)

seantalts replied:
To answer my 2nd question, it does still run with `mpirun` even if they aren't mpi tests.

syclik commented on runTests.py:
```diff
@@ -97,10 +97,12 @@ def makeTest(name, j):
     xml = mungeName(name).replace(winsfx, "")
     command = '%s --gtest_output="xml:%s.xml"' % (executable, xml)
+    if mpi and "mpi_" in name:
+        command = "mpirun -np 2 " + command
```
should we include a parameter to change that `2` to something that's specified by the user?

wds15 replied:
making that configurable would be nice, but I would also be fine to do with this logic:

- at least 2
- if -j#CPUs is given and #CPUs is >2, then just take that number

We want at least two processes, but possibly more.

seantalts replied:
Okay, will do @wds15 suggestion here

wds15 commented:
@seantalts Could you look into these remaining requests from @syclik ? These are wrt to the Jenkins config and the runTests.py. Thanks!

seantalts commented:
Woops, yes will do that today! Exciting times folks
On Wed, Mar 14, 2018 at 04:11 wds15 <notifications@github.com> wrote:

> @seantalts <https://github.com/seantalts> Could you look into these
> remaining requests from @syclik <https://github.com/syclik> ? These are
> wrt to the Jenkins config and the runTests.py. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-372937365>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7JVHZQFQ2XQUCD3PXu29_IFkKVgpks5teNDCgaJpZM4Sa35S>
> .
>

seantalts commented:
Okay, updated to do as you guys suggested. Please ping me again if I can help more!

wds15 commented:
Jenkins appears to be unstable. I restarted a few times and I end up with the "!" in all boxes / Jenkins deeming results as "unstable".

seantalts commented:
Unstable means that tests are failing, though it seems to have encountered some kind of Jenkins bug while trying to parse the test XMLs. There are also some new compiler warnings. Here are the warnings from the full console log:
```
[Unit with MPI] In file included from ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:14:0,
[Unit with MPI]                  from ./stan/math/rev/mat.hpp:57,
[Unit with MPI]                  from ./stan/math.hpp:4,
[Unit with MPI]                  from test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:1:
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, f1(std::_Placeholder<1>, double, std::reference_wrapper<std::basic_ostream<char> >)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:151:22:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f1; T_param = double; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:49:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
[Unit with MPI]    int num_function_evaluations;
[Unit with MPI]        ^
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, double (*(f1, std::_Placeholder<1>, stan::math::var, stan::math::var, std::reference_wrapper<std::basic_ostream<char> >))(const f1&, double, const stan::math::var&, const stan::math::var&, std::basic_ostream<char>&)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:167:35:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f1; T_param = double; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:49:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, f2(std::_Placeholder<1>, std::vector<double>, std::reference_wrapper<std::basic_ostream<char> >)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:151:22:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f2; T_param = std::vector<double>; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:53:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, double (*(f2, std::_Placeholder<1>, std::vector<stan::math::var>, stan::math::var, std::reference_wrapper<std::basic_ostream<char> >))(const f2&, double, const std::vector<stan::math::var>&, const stan::math::var&, std::basic_ostream<char>&)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:167:35:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f2; T_param = std::vector<double>; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:53:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, f3(std::_Placeholder<1>, Eigen::Matrix<double, -1, 1>, std::reference_wrapper<std::basic_ostream<char> >)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:151:22:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f3; T_param = Eigen::Matrix<double, -1, 1>; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:62:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, double (*(f3, std::_Placeholder<1>, Eigen::Matrix<stan::math::var, -1, 1>, stan::math::var, std::reference_wrapper<std::basic_ostream<char> >))(const f3&, double, const Eigen::Matrix<stan::math::var, -1, 1>&, const stan::math::var&, std::basic_ostream<char>&)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:167:35:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f3; T_param = Eigen::Matrix<double, -1, 1>; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:62:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, f4(std::_Placeholder<1>, Eigen::Matrix<double, 1, -1>, std::reference_wrapper<std::basic_ostream<char> >)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:151:22:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f4; T_param = Eigen::Matrix<double, 1, -1>; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:70:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp: In instantiation of ‘double stan::math::de_integrator(const F&, double, double, double, double) [with F = std::_Bind_result<double, double (*(f4, std::_Placeholder<1>, Eigen::Matrix<stan::math::var, 1, -1>, stan::math::var, std::reference_wrapper<std::basic_ostream<char> >))(const f4&, double, const Eigen::Matrix<stan::math::var, 1, -1>&, const stan::math::var&, std::basic_ostream<char>&)>]’:
[Unit with MPI] ./stan/math/rev/mat/functor/integrate_1d_tsc.hpp:167:35:   required from ‘typename stan::scalar_type<T2>::type stan::math::integrate_1d_tsc(const F&, double, double, const T_param&, std::ostream&, double, double) [with F = f4; T_param = Eigen::Matrix<double, 1, -1>; typename stan::scalar_type<T2>::type = double; std::ostream = std::basic_ostream<char>]’
[Unit with MPI] test/unit/math/prim/arr/functor/integrate_1d_tsc_test.cpp:70:3:   required from here
[Unit with MPI] ./stan/math/rev/mat/functor/de_integrator.hpp:65:7: warning: variable ‘num_function_evaluations’ set but not used [-Wunused-but-set-variable]
```

seantalts commented:
I'll dig into this more today, seems like something I should be able to help with.

wds15 commented:
Thanks!

seantalts commented:
Oh, I think this is probably from using the mpi compiler, which is grabbing the wrong compiler from the system.

wds15 commented:
Looks like this went wrong.

It would be best if mpich is compiled itself with the compiler which we want to use for testing. Using a different compiler than the one which was used to build mpich is not really recommended. This is due to mpich figuring out the bit representations of all types when its build. I am not sure if changing the compiler is a good idea then.

seantalts commented:
Okay, I think everything is building correctly now, but when the tests are executing, they're doing so with two executables running in parallel writing to the same XML file, corrupting it. There's some information about workarounds [here](https://github.com/google/googletest/issues/822). @wds15 want to take a look at that?

wds15 commented:
Are we currently running any test with mpirun? So even those which are not truly mpi tests? If yes, could we change that such that only MPI tests are executed with mpirun?

Alternativley we could run everything with mpirun -np 1 for the non-MPI tests.

If this is not an option (or not a solution), then we need to dig further.

For the true MPI tests I am already messing with the output such that the real MPI tests should be ok...at least I would hope that.
I am not on a computer until Thursday and can only try these ideas then.

syclik commented:
btw, I'm really glad we're tackling these issues now before we get this
into the code base. this type of config wreaks havoc on our users.

On Mon, Mar 19, 2018 at 1:58 PM, wds15 <notifications@github.com> wrote:

> Are we currently running any test with mpirun? So even those which are not
> truly mpi tests? If yes, could we change that such that only MPI tests are
> executed with mpirun?
>
> Alternativley we could run everything with mpirun -np 1 for the non-MPI
> tests.
>
> If this is not an option (or not a solution), then we need to dig further.
>
> For the true MPI tests I am already messing with the output such that the
> real MPI tests should be ok...at least I would hope that.
> I am not on a computer until Thursday and can only try these ideas then.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-374307654>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_FyzAttMIc6t_Pt8JmUupSEps8Cw-ks5tf_G0gaJpZM4Sa35S>
> .
>

seantalts commented:
@syclik had requested we run all unit tests again with the MPI
configuration.
On Mon, Mar 19, 2018 at 14:00 Daniel Lee <notifications@github.com> wrote:

> btw, I'm really glad we're tackling these issues now before we get this
> into the code base. this type of config wreaks havoc on our users.
>
> On Mon, Mar 19, 2018 at 1:58 PM, wds15 <notifications@github.com> wrote:
>
> > Are we currently running any test with mpirun? So even those which are
> not
> > truly mpi tests? If yes, could we change that such that only MPI tests
> are
> > executed with mpirun?
> >
> > Alternativley we could run everything with mpirun -np 1 for the non-MPI
> > tests.
> >
> > If this is not an option (or not a solution), then we need to dig
> further.
> >
> > For the true MPI tests I am already messing with the output such that the
> > real MPI tests should be ok...at least I would hope that.
> > I am not on a computer until Thursday and can only try these ideas then.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/stan-dev/math/pull/783#issuecomment-374307654>, or
> mute
> > the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAZ_FyzAttMIc6t_Pt8JmUupSEps8Cw-ks5tf_G0gaJpZM4Sa35S
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/783#issuecomment-374308566>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAxJ7FBKIcCCTYhy3XwGRqU3khO5v1mqks5tf_JXgaJpZM4Sa35S>
> .
>

wds15 commented:
We would still run things with mpirun, just with a single process. Is that ok for @syclik ? Otherwise we have to adjust the gtest main cpp file to deal with multiple processes. If we do that, then I would basically suppress execution of the tests for any test other than those running with rank != 0.

seantalts commented:
While we're waiting for his answer, I committed a change that I think will do as you're suggesting.

syclik commented:
@wds15 and @seantalts, it looks like it's tricky gtest to run with multiple threads. I think @seantalts already linked to google/googletest#822.

Below are questions that I haven't thought through and think we should answer:

- when we run a unit test, when do we want to run it with `mpirun`? (why do we run it with `mpirun` when it's an "mpi test"?)
- is there a way to run mpi properly without invoking `mpirun`?
- is there a way to install mpi without installing `mpirun`? What do we do then?
- where should we put this documentation within the Math library? Some reasonable options: Math wiki, makefile help, doxygen doc.

And just to verify, the unit tests are run with mpi and without mpi now? (That's what I think it does, but just wanted to verify.)

seantalts commented:
We've been talking about some of these already, in this thread and elsewhere. Re: doc - I will add something to runTests.py to print out if mpi is enabled and `mpirun` is not found and link it to this wiki page: https://github.com/stan-dev/stan/wiki/Parallelism-using-MPI-in-Stan (where I think we should add a "how to run the tests" section, and maybe update / rewrite for users instead of a design doc (not it). I think we can have instructions on what needs to be installed and perhaps even links to installation instructions on the wiki, and instructions on how to run the mpi tests once everything is installed correctly.

wds15 commented:
I can give more detail later, but mpirun is always part of the base mpi installation. So we can assume its there...otherwise the mpi installation is messed up, but this is a pre—requesite of us.

wds15 commented:
Let me go through your questions in order:

1. If you want to run things as close as possible as under MPI conditions, then we always want run tests with mpirun - even those non-MPI tests. However, the non-MPI tests have to be run with only a single thread as these are not capable to deal with multiple threads.

2. I do not know if you can run MPI properly without it, but it is always available whenever there is a base MPI installation.

3. I have never seen an MPI installation without mpirun, but I can only speak for the systems I have dealt with.

4. For stan-math we should probably have MPI unit test specifics in the wiki with a bit more detail and the bare basic info in help printouts, just as @seantalts suggests.

For the record: My understanding of mpirun is that it fires up as many processes as it is supposed to on a given machine. The specifcs of what is started where depends on its command line arguments which includes an optional hostfile which contains the info how many processes are started on which machine in a bigger cluster. Remember that mpirun is usually started from a queing system which assigns these ressources in full automation. On a given machine mpirun then starts the number of processes and codes in the environment of each process all the necessary info which the process needs to know in order to work with all the other MPI processes (which are on the same machine and/or on other machines). I think that you can get by without mpirun, but that is not recommended by any means as I understood it. So mpirun sets up everything needed to make MPI work (including how communication takes place). As MPI is very flexible in how communication is performed there are a few bells and whistles, but for unit testing we only need to deal with the default case and make sure that things work in this case.

What is left now to do, in order to merge?

wds15 commented:
@syclik just let me know where what documentation should go to. I am also happy to make a first draft as I envision.

BTW, where will we store extra documentation for the GPU stuff? We should probably make this in a consistent manner.

wds15 commented:
Bump. I put down a lot of documentation in our wiki:

https://github.com/stan-dev/stan/wiki/Parallelism-using-MPI-in-Stan

@syclik Is this along the lines you thought of? It's easy for me to overlook stuff as I am used to it, but I tried to have a comprehensive text on the matter.

syclik commented:
Thanks for putting up the doc. A couple generic things:
1. Can you add a link to the wiki on the original post? It would help to look there instead of searching through the comments to find it.
2. Since this is just for control of that Math library, could you move the wiki to Math? I don't think this applies to Stan or CmdStan yet.

Specific things for the doc:
1. We're missing an installation guide to MPI. I understand how it's written, but the audience here is going to be people who want to run MPI and need help setting up first. That's missing.
2. It needs to clearly state that it will not work for Windows. This PR doesn't allow that to work. We should explicitly state that.
3. More detail? I don't know how to "Ensure that a base MPI installation is available and accessible on the system" -- what does that mean?
4. The "How to run MPI tests..." section is really confusing. I don't think I understand how to run them from reading that. Perhaps a code block with an example would help.

wds15 commented:
Your first two generic things:

1. It's not quite clear to me what you are referring to. Which post should be changed with a link to the wiki?
2. Makes sense. Moved to `stan-math` wiki, [now here](https://github.com/stan-dev/math/wiki/MPI-Parallelism).

Specifics:

1. No, the MPI facility is tailored at (Linux) clusters where we can expect that a MPI installation is already available. We always said so (my understanding). I am making this more clear in the doc now, but I also give pointers as to how a "normal" user can use the facility. However, on a single machine users should really just use the threading stuff.
2. Ok, Windows is now mentioned. The only reason to disable windows is that we are unable to deploy MPI on windows. If a clever user manages to get MPI work on windows, then he can just use it. This is why I felt to just not mention Windows in the first place.
3. Ensuring that MPI works is described in the text by making reference to the `boost.mpi` documentation. I have moved this sentence to an earlier place. As we target systems where MPI already works I think its fine to point to other places where such details are described.
4. I refined this "how to run the MPI tests" section and now also give a code block. This overall block is more than just a "how to run". It also details what is happening. If it were just a how to run, then we could just say "use runTests.py" and the rest is done in full automation.

bob-carpenter commented:
Thanks for building out the doc here.  The install doc and smoothness is going to be important for getting people spun up on this.

I agree that we don't want to maintain MPI install instructions.  It's a grey area here between something like MPI and something like R for RStan.  

What we want to avoid on Windows is ordinary users trying to turn on MPI and having it not work.  We should make it clear in the doc that MPI doesn't work on Windows.  We should also disable it if possible on Windows so that a Windows user will get an error trying to turn MPI on.  Then, we can also ask for help making it work for Windows and provide the hint that they'll need to turn off whatever we do to disable Windows.

syclik commented:
@wds15, I'm typing on my phone, so please forgive briefness. 

- The original post I think should change is here; I'm not sure how to reference it properly: https://github.com/stan-dev/math/pull/783#issue-172661790
- thanks for moving the page to the math wiki. 
- I wasn't trying to say we need to write every detail on how to install. But I'd still like to be friendly to our users. I don't think most of our users are Linux cluster users. That's probably where we're just not on the same page. Even the devs are trying to test this locally. With that in mind, it'd help to just say something like: "here are some resources on how to install MPI."
- re: windows, doesn't this PR explicitly make it not run on Windows?
- thanks for cleaning up the doc a lot. It's a lot easier to read.

bob-carpenter commented:
Links to resources are good, especially if they're maintained.  But I wouldn't let worry about maintenance prevent us from including some.

Going forward, we'll be recommending multi-process (MPI) for clusters of multiple machines and multi-threading (C++11 threads) for multi-core single machines.  GPUs will be usable either way, though we have to worry about GPU scheduling with multiple threads or processes on the same machine.

Windows users trying to specify MPI should get an error rather than a fallback to a serial implementation.

> On Mar 31, 2018, at 7:36 PM, Daniel Lee <notifications@github.com> wrote:
> 
> @wds15, I'm typing on my phone, so please forgive briefness.
> 
> 	• The original post I think should change is here; I'm not sure how to reference it properly: #783 (comment)
> 	• thanks for moving the page to the math wiki.
> 	• I wasn't trying to say we need to write every detail on how to install. But I'd still like to be friendly to our users. I don't think most of our users are Linux cluster users. That's probably where we're just not on the same page. Even the devs are trying to test this locally. With that in mind, it'd help to just say something like: "here are some resources on how to install MPI."
> 	• re: windows, doesn't this PR explicitly make it not run on Windows?
> 	• thanks for cleaning up the doc a lot. It's a lot easier to read.
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

syclik commented:
@wds15, I updated the documentation. See it here:
https://github.com/stan-dev/math/wiki/MPI-Parallelism

I've written it so that someone can find relevant information on:
- what it's doing
- how to find resources on installation
- how to set up the Math library
- how to run.

Hopefully that change to documentation makes sense. I don't think it's complete and it's not that much different in total content, but it makes a lot of difference trying to follow along as someone that's not familiar with MPI.

syclik commented:
I just merged develop and pushed. Once it passes tests, please merge.

wds15 commented:
Cool! Thanks much. Next time I will make more use of bold headers, etc.
